### ISSUE TITLE
Feature: Trusted stream for webcam input

### ISSUE THREAD

**Issue by tasket** [QubesOS/qubes-issues#2079]
Create a uni-directional (flowing from non-networked webcam vm) webcam stream that is sanitized in much the same way as static images or pdfs. Or, more accurately, similar to microphone input.

This could show up as a device Attach/Detach option in appvm context menus alongside the microphone, and would enable videoconferencing sessions with trusted services and peers. Ultimately, this feature would help enable an important PC use case: Telecommuting and conferencing with video and window streams.

Technical: Even though re-compression of the stream would have to occur (probably in the vm running the conferencing app) this should be feasible today if resolution and frame rate are kept at a moderate level. Bounds-checking will have an added dimension: Frame rate. If the stream is pushing more than a certain percentage over a set frame rate for the last n seconds, receiving vm will drop excess frames.

**Comment by marmarek** [comment]
In theory it is good idea. In practice it probably will fail on performance. In 3.2 we've got USB passthrough. When used for webcam (raw USB packets, most likely compressed) it already results in very high CPU usage. If some additional processing would be added (for example converting to "simple representation"), almost for sure it would be too slow.

But if that performance problem would be solved somehow, it would be much better for video conferences than full USB passthrough.

This wouldn't solve privacy problem completely (`sys-usb` still would be able to sniff all the video stream, and later leak it through some other USB device). But this problem is very hard to solve (if possible at all) with the current hardware (lack of encrypted videostream from webcams, lack of secure way of isolating single USB device).

**Comment by ttasket** [comment]
Yes, today's announcement is what gave me the idea because the security risk still exists with USB passthrough.

I've seen how re-encoding streams in real time can work even in a laptop appvm. It may drop frames, but its not at all terrible. Firefox extensions that can screen-grab video players are a good example:
https://www.downloadhelper.net

It may even be the case that PCI passthrough for GPUs becomes more stable and thus enables a sanitized video stream to be efficiently re-encoded in real time.

**Comment by ttasket** [comment]
Discussion thread at [qubes-devel.](https://groups.google.com/d/msgid/qubes-devel/57654B5C.7000204%40openmailbox.org)

**Comment by nothingmuch** [comment]
I found this:

https://github.com/umlaeute/v4l2loopback

For my use case, work hangouts in google chrome, this might be enough because it uses V4L2 (https://src.chromium.org/viewvc/chrome/trunk/src/media/video/capture/linux/video_capture_device_linux.cc)

At least in principle it should be possible to capture video input in sys-usb (without even demuxing, let alone decoding), and expose it as an opaque data stream to the appvm instead of as a USB device, and feed it into the loopback video driver in the appvm's userspace. It might even have less overhead than USB forwarding, if it uses buffered IO and avoids treating the data stream as anything more than binary data.

Note, I'm not really familiar with V4L so please assume I have no clue what I'm talking about.

**Comment by stiell** [comment]
USB passthrough did not work for me, so I've tried an approach similar to that described by @nothingmuch and can confirm that it works. I used ffmpeg on sys-usb to feed webcam input to a Qubes RPC pipe going to the appvm, where it would be decoded by ffmpeg and fed to a v4l2loopback device.

**Comment by marmarek** [comment]
@stiell do you care to share the scrips?

**Comment by stiell** [comment]
I accidentally posted the instructions in a different issue: #2594

**Comment by tfm1** [comment]
Thanks for sharing those instructions!

On my system using ffmpe-qrexec stream is roughly the same overhead as using USB forwarding. ffmepg does seem to have a slight latency though. But it's barely noticeable and I doubt it will be an issue for video conferencing.

Issues remaining are: (1) Need to use HVM to have the kernel module. I believe this could be fixed by providing the right kernel through dom0, right?
(2) The webcam will constantly stream while connected to a VM. This will result in a slight overhead (like 25% CPU usage on a 4 CPU hardware) and the webcam's LED will be on.

That last point I'm actually starting to see as a feature. The webcam's LED indicator will tell you that it is connected to a VM as a warning that the VM could start streaming video any time. You usually wouldn't want to have the stream connected for longer than you really use it. With USB forwarding there is a bigger risk of forgetting that your camera is connected to some VM.

**Comment by ben-grande** [comment]
Solved by https://github.com/qubesos/qubes-video-companion
